api/metrics-api-jaxrs/src/main/java/org/hawkular/metrics/api/jaxrs/log/HttpErrorLoggerExtension.java

```diff
@@ -0,0 +1,130 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.hawkular.metrics.api.jaxrs.log;
+
+import java.nio.ByteBuffer;
+import java.nio.charset.Charset;
+import java.util.regex.Pattern;
+
+import javax.servlet.ServletContext;
+
+import org.xnio.channels.StreamSourceChannel;
+
+import io.undertow.connector.PooledByteBuffer;
+import io.undertow.server.ExchangeCompletionListener;
+import io.undertow.server.HttpHandler;
+import io.undertow.server.HttpServerExchange;
+import io.undertow.servlet.ServletExtension;
+import io.undertow.servlet.api.DeploymentInfo;
+import io.undertow.util.AttachmentKey;
+import io.undertow.util.HttpString;
+import io.undertow.util.Methods;
+import io.undertow.util.StatusCodes;
+
+
+public class HttpErrorLoggerExtension implements ServletExtension {
+
+    private static final RestLogger log = RestLogging.getRestLogger(HttpErrorLoggerExtension.class);
+    private HttpErrorHandler httpErrorHandler;
+    private static final AttachmentKey<String> HTTP_ERROR_LOGGER_KEY = AttachmentKey.create(String.class);
+
+    Pattern postQueryPattern;
+
+    class HttpErrorExchangeCompleteListener implements ExchangeCompletionListener {
+        @Override
+        public void exchangeEvent(HttpServerExchange exchange, ExchangeCompletionListener.NextListener nextListener) {
+            int httpStatusCode = exchange.getStatusCode();
+            if (httpStatusCode >= StatusCodes.BAD_REQUEST) {
+                final String path;
+                final String query = exchange.getQueryString();
+                if (!query.isEmpty()) {
+                    path = exchange.getRequestPath() + "?" + query;
+                } else {
+                    path = exchange.getRequestPath();
+                }
+                HttpString method = exchange.getRequestMethod();
+                String requestBody = exchange.getAttachment(HTTP_ERROR_LOGGER_KEY);
+                if (requestBody==null) {
+                    log.errorf("Service %s %s fails with HTTP code: %d", method, path,
+                            httpStatusCode);
+                    exchange.removeAttachment(HTTP_ERROR_LOGGER_KEY);
+                } else {
+                    log.errorf("Service %s %s fails with HTTP code: %d, body: %s", method, path,
+                            httpStatusCode, requestBody);
+                }
+
+
+            }
+            nextListener.proceed();
+        }
+    }
+
+    class HttpErrorHandler implements HttpHandler {
+
+        private HttpHandler next;
+        Pattern postQueryPattern;
+
+        public HttpErrorHandler(HttpHandler next, Pattern postQueryPattern) {
+            this.next = next;
+            this.postQueryPattern = postQueryPattern;
+        }
+
+        private void getRequestBody(HttpServerExchange exchange) {
```

Comment:
This method allocates huge amount of stuff, once the bytes from the stream to the memory, then a byte array and then a String from that byte array.

Reply:
Yeah I know, indeed this way of reading the request body doesn't work very well :(